File: src/main/webapp/src/app/account/model/account-api.ts
import {Profile} from "../../profile/model/profile-api";

export interface Account {
  id: number,
  externalId: string | undefined,
  name: string,
  iconUrl: string,

  registrationDate: string,
  lastLoginDate: string | undefined,

  serviceAccount: boolean,
  idpLinked: boolean,
}

export interface AccountInfo {
  id: number,
  name: string,
  iconUrl: string,
  email: string,
  firstName: string,
  lastName: string,
}

export interface AccountValidationResult {
  created: boolean,
  account: Account,
  profile: Profile,
  info: AccountInfo
}

export class AccountChangeRequest {
  constructor(
    public name: string,
    public iconUrl: string,
    public externalId: string | undefined
  ) {
  }
}

File: backend/src/main/kotlin/de/sambalmueslie/openevent/api/AccountAPI.kt
package de.sambalmueslie.openevent.api

import de.sambalmueslie.openevent.core.account.api.*
import io.micronaut.data.model.Page
import io.micronaut.data.model.Pageable
import io.micronaut.security.authentication.Authentication

interface AccountAPI : CrudAPI<Long, Account, AccountChangeRequest> {
    companion object {
        const val PERMISSION_READ = "openevent.account.read"
        const val PERMISSION_WRITE = "openevent.account.write"
        const val PERMISSION_ADMIN = "openevent.account.admin"
        const val PERMISSION_MODERATOR = "openevent.account.mod"
    }

    fun findByExternalId(auth: Authentication, externalId: String): Account?
    fun findByName(auth: Authentication, name: String, pageable: Pageable): Page<Account>
    fun findByEmail(auth: Authentication, email: String): Account?
    fun validate(auth: Authentication, lang: String = ""): AccountValidationResult

    fun getProfile(auth: Authentication): Profile?
    fun getPreferences(auth: Authentication): Preferences?

    fun setup(auth: Authentication, request: AccountSetupRequest): AccountInfo?
    fun update(auth: Authentication, id: Long, request: AccountSetupRequest): AccountInfo?
}
